workflow-core/src/main/java/com/squareup/workflow/WorkflowIdentifier.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2020 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow
+
+import okio.BufferedSink
+import okio.BufferedSource
+import okio.IOException
+import kotlin.LazyThreadSafetyMode.PUBLICATION
+import kotlin.reflect.KClass
+
+/**
+ * Represents a [Workflow]'s "identity" and is used by the runtime to determine whether a workflow
+ * is the same as a workflow as was rendered in a previous render pass, in which case its state
+ * should be re-used, or if it's a new workflow and needs to be started.
+ *
+ * The identity consists primarily of the workflow's concrete type (i.e. the class that implements
+ * the [Workflow] interface). Two workflows of the same concrete type are considered identical.
+ *
+ * However, if a workflow class implements [ImpostorWorkflow], the identifier will also include
+ * that workflow's [ImpostorWorkflow.realIdentifier].
+ *
+ * @constructor
+ * The workflow's type must be passed _either_ as a [KClass] or a string, but only one must be
+ * non-null. The [KClass] should be passed in if it's available, to avoid reflection calls upfront,
+ * but the class's name may be passed instead if the [KClass] is not available, eg. when restoring
+ * this class from its serialized form.
+ *
+ * The workflow type must be final, since a non-final class or interface means two different
+ * workflow types could be considered identical.
+ *
+ * @param type The [KClass] of the [Workflow] this identifier identifies.
+ * @param typeString The fully-qualified name of the [KClass] of the workflow this identifier
+ * identifies.
+ * @param proxiedIdentifier An optional identifier from [ImpostorWorkflow.realIdentifier] that will
+ * be used to further narrow the scope of this identifier.
+ */
+@ExperimentalWorkflow
+class WorkflowIdentifier internal constructor(
+  private val type: KClass<out Workflow<*, *, *>>? = null,
+  typeString: String? = null,
+  private val proxiedIdentifier: WorkflowIdentifier?
+) {
+  init {
+    require((type == null) xor (typeString == null)) {
+      "Either type or type string must be passed, not both (type=$type, typeString=$typeString)"
+    }
+  }
+
+  /**
+   * The fully-qualified name of [type]. Computed lazily.
+   */
+  private val typeString: String by lazy(PUBLICATION) { typeString ?: type!!.java.name }
```

Comment:
I don't think this laziness is worth it, but highlights an issue with this approach. Whereas the old code would look up the actual classes for each restored workflow, this code just stores the string instead. Since comparisons with string identifiers require turning classes into names, this means that any time a tree is restored from a snapshot, all workflow classes will have their names reflected on every render pass. I haven't benchmarked this, but it seems less than ideal.

Reply:
We should probably go back to just looking up classes for now, until we measure this change.

Reply:
Went back to eagerly resolving classes (i.e. the old behavior). If we want to change this later, it should be a separate change anyway.